### PR TITLE
feat: Surface artifacts through sidecar container logs. 

### PR DESCRIPTION
**Pull request by ericzzzzzzz**
# Changes
 - This Pr uses the similar approach as https://github.com/tektoncd/pipeline/pull/7414 to surface large artifacts to TaskRun.status.steps. 
 -  The flag `result-from:sidecar-logs`  is used to  guard this feature, it is also the same as surfacing step results through sidecar logs. 
 - As we don't have a API to allow users to declare if a step is going to produce artifacts, we check if a step env, command, args, script contains `artifacts path` to determine if we should create an artifact sidecar container should be created for a task. 

# Fixes: #7869      

# Submitter Checklist

As the author of this PR, please check off the items in this checklist:

- [ ] Has [Docs](https://github.com/tektoncd/community/blob/main/standards.md#docs) if any changes are user facing, including updates to minimum requirements e.g. Kubernetes version bumps
- [x] Has [Tests](https://github.com/tektoncd/community/blob/main/standards.md#tests) included if any functionality added or changed
- [ ] [pre-commit](https://github.com/tektoncd/pipeline/blob/main/DEVELOPMENT.md#install-tools) Passed
- [x] Follows the [commit message standard](https://github.com/tektoncd/community/blob/main/standards.md#commits)
- [x] Meets the [Tekton contributor standards](https://github.com/tektoncd/community/blob/main/standards.md) (including functionality, content, code)
- [x] Has a kind label. You can add one by adding a comment on this PR that contains `/kind <type>`. Valid types are bug, cleanup, design, documentation, feature, flake, misc, question, tep
- [x] Release notes block below has been updated with any user facing changes (API changes, bug fixes, changes requiring upgrade notices or deprecation warnings). See some examples of [good release notes](https://github.com/tektoncd/community/blob/main/standards.md#good-release-notes).
- [ ] Release notes contains the string "action required" if the change requires additional action from users switching to the new release

# Release Notes

```release-note
Surface artifacts through sidecar container logs.
```

### CODE REVIEW

**Comment by ericzzzzzzz** [comment]
/hold 

needs to discuss if we need api changes

File: internal/sidecarlogartifacts/sidecarlogartifacts.go [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+	info, err := os.Stat(filename)
+	if os.IsNotExist(err) {
+		return false, nil
+	} else if err != nil {
+		return false, fmt.Errorf("error checking for file existence %w", err)
+	}
+	return !info.IsDir(), nil
+}
+
+func waitForStepsToFinish(runDir string) error {
```
Comment (chitrangpatel) [inline internal/sidecarlogartifacts/sidecarlogartifacts.go:44]: any way to rely on result's code or refactoring this such that its not duplicated between results and artifacts? 

The same for all other functions.

Comment (chitrangpatel) [reply internal/sidecarlogartifacts/sidecarlogartifacts.go:44]: i.e. have a shared library that both the images can rely on.

**Comment by ericzzzzzzz** [comment]
> At a high level, I wonder if we need two separate sidecars (one for artifacts and one for results) or if we can have the same sidecar log them both? Obviously, we'll need to distinguish between artifacts and results in that case.

That's a great point to consider.
 - For two sidecar approaches, 
    Pros:
       - Each sidecar would have a dedicated purpose, making the system easier to understand.
    Cons:
       - duplicate code
       - one more container creation 
 - For one sidecar approach, this also seems reasonable. 
   - Cons:
        - we will need to treat artifacts as a special `Result`, we're already doing this with termination message approach, so that isn't a problem
   - Pros:
      - we can also take advantage of the existing code in resultsidecar log.
      - unify result extractions logic when setting task/steps status seems reasonable 

Anyway, I'll explore the possibility of a single sidecar to see which approach makes more sense, we can chat about that tomorrow. : )

**Review by JeromeJu — COMMENTED** [review]
Thanks @ericzzzzzzz  ! This generally looks good. Just some nit picking and do we have any failure test cases for the coverage?

File: internal/sidecarlogresults/sidecarlogresults.go [hunk 2]
```diff
@@ -260,3 +298,19 @@ func parseResults(resultBytes []byte, maxResultLimit int) (result.RunResult, err
...
+	if err := json.Unmarshal(fileContent, &as); err != nil {
+		return as, fmt.Errorf("invalid artifacts : %w", err)
+	}
+	return as, nil
+}
+
+func extractArtifactsFromFile(filename string) (v1.Artifacts, error) {
+	b, err := os.ReadFile(filename)
+	if err != nil {
+		return v1.Artifacts{}, err
```
[the remark this answers is not in the document]
Comment (ericzzzzzzz) [reply internal/sidecarlogresults/sidecarlogresults.go]: added

File: pkg/pod/pod.go [hunk 3]
```diff
@@ -734,3 +749,31 @@ func usesWindows(tr *v1.TaskRun) bool {
 	osSelector := tr.Spec.PodTemplate.NodeSelector[osSelectorLabel]
 	return osSelector == "windows"
 }
+
+func artifactsPathReferenced(steps []v1.Step) bool {
+	// `$(step.artifacts.path)` in  taskRun.Spec.TaskSpec.Steps and `taskSpec.steps` are substituted when building the pod while when setting status for taskRun
+	// neither of them is substituted, so we need two forms to check if artifactsPath is referenced in steps.
+	unresolvedPath := "$(step.artifacts.path)"
```
[the remark this answers is not in the document]
Comment (ericzzzzzzz) [reply pkg/pod/pod.go]: updated.

File: pkg/entrypoint/entrypointer.go [hunk 4]
```diff
@@ -292,15 +292,15 @@ func (e Entrypointer) Go() error {
 	return err
 }
 
-func readArtifacts(fp string) ([]result.RunResult, error) {
+func readArtifacts(fp string, resultType result.ResultType) ([]result.RunResult, error) {
 	file, err := os.ReadFile(fp)
 	if os.IsNotExist(err) {
 		return []result.RunResult{}, nil
```
[the remark this answers is not in the document]
Comment (ericzzzzzzz) [reply pkg/entrypoint/entrypointer.go]: Yes! This is intentional, for two reasons 
 - We don't have artifact api at the moment and when I initially implemented this, I didn't find a way to know if a step should produce artifacts -- if file is supposed to exist. For this, we can check if step references steps.artifact.path to know if this file is supposed to exist and I also updated the corresponding change when creating result sidecar container. So this is not an issue anymore. 
 -  This is consistent with surfacing results through sidecarlog, it doesn't error out either when result file not found. https://github.com/tektoncd/pipeline/commit/30540fc9a219097b468d2d4ca0e1f4171520ff4e

**Comment by ericzzzzzzz** [comment]
> Thanks @ericzzzzzzz ! This generally looks good. Just some nit picking and do we have any failure test cases for the coverage?

Thanks for the review! I added some tests,  the some failure cases are related to file operations and we don't have established pattern to mock those, others are related to json marshal  operation, it shouldn't happen as we are passing a struct with all tagged public fields

**Review by JeromeJu — APPROVED** [review]

**Comment by JeromeJu** [comment]
> > Thanks @ericzzzzzzz ! This generally looks good. Just some nit picking and do we have any failure test cases for the coverage?
> 
> Thanks for the review! I added some tests, the some failure cases are related to file operations and we don't have established pattern to mock those, others are related to json marshal operation, it shouldn't happen as we are passing a struct with all tagged public fields

Understood - then maybe wrapping the errors with the appropriate context might be the way to go.

**Comment by chitrangpatel** [comment]
I'm merging this PR since it is basically taking the exact same approach as for larger results to extract artifacts. This has been open for a while and blocking other work. The same functionality has already been implemented for surfacing the information via TerminationMessages. There is nothing novel here.

/lgtm

**Pull request merged**